doc/design.md

```diff
@@ -0,0 +1,1268 @@
+
+
+### <a name="design-details-threading"></a>Operations are threaded using OpenMP
+
+##### Rationale
+
+Threading using OpenMP in a fork-join style manner is not the most efficient one.
+As part of the prototyping process we considered other solutions, in particular [dask](https://dask.org/).
+Dask could solve many problems such as overcoming the limitations of the fork-join approach as providing a unified solution for threading and multi-process parallelisation such as MPI.
+However, the prototyping experience indicated that it is not performant enough to completely replace threading for our typical workloads.
+For that reason, and for maximum simplicity, optimal performance requires multi-threading.
+
+##### Implementation
+
+In most cases a simple `#pragma omp parallel for` could do the job.
+We should probably limit the thread count depending on the size of data and the operation in question.
+For small data the threading overhead may be larger than the gains.
+For large data, many cases such as `operator+` are heavily limited by memory bandwidth and using more cores than necessary for reaching the peak bandwidth would be wasteful.
+
+
+### <a name="design-details-no-magic"></a>No magic
+
+##### Rationale
+
+Many algorithms in Mantid do "magic" under the hood:
+- Convert inputs if required. *Example: Conversion from point data in `Algorithms::ConvertUnits` and `Algorithms::Rebin`.*
+- Tolerate partially bad input data. *Example: Replacing special values in `Algorithms::SumSpectra`.*
+- Support doing two different things that are actually different. *Example: `Algorithms::Rebin` can rebin histogram data, bin event data, or set new bin edges in a workspace.*
+- Combine multiple processing steps. *Examples: Aligning bins in `Algorithms::ConvertUnits`. Loading and saving calibration files in `Algorithms::MaskDetectorsIf`.*
+
+Support for such features complicates the implementation and makes reasoning about code harder.
+It should therefore not be done in this library.
+- Operations should simply fail as early as possible.
+- Inputs are required to be cleaned up and converted to the correct format explicitly.
+- Processing steps should be kept as separate operations.
+
+##### Note
+
+There is nothing wrong with "magic" in higher-level operations.
+High level algorithms that simply work no matter what is thrown at them are likely still a requirement.
+We are merely banning it from *this* library, i.e., the lowest-level building block.
+
+
+### <a name="design-details-no-logging"></a>No logging
+
+##### Rationale
+
+As a corollary from the "no magic" strategy above, there will be very few occasions that require log messages.
+Logging can be implemented in higher-level algorithms, or potentially in some Python wrappers.
+Adding logging functionality would increase complexity and add extra dependencies.
+
+##### Workaround
+
+If required we could log to `stdout` using `std::cout`, which can be redirected and filtered.
```

Comment:
I would like to avoid introducing a dependency on `Poco`. But as I am pointing out here, it is possible to redirect `std::out` programmatically, i.e., if the `dataset` library is embedded into something higher-level like Mantid you should be able to redirect it into our current Poco logging framework.

Reply:
At the `dataset` level perhaps do something similar to other libraries where you can install a message handler and then don't use `cout` directly. Mantid could then simply install a handler that went to the logging framework.

Reply:
Sounds like a good suggestion, do you have a reference where this has been done before?

Reply:
Qt has [`qInstallMessageHandler`](http://doc.qt.io/qt-5/qtglobal.html#qInstallMessageHandler) which is basic the type of thing I was suggesting.

Reply:
VTK had a recent discussion about adding logging. Thought it might be relevant here

https://public.kitware.com/pipermail/vtk-developers/2018-August/036256.html